com.unity.netcode.gameobjects/Runtime/Serialization/BitWriter.cs

```diff
@@ -0,0 +1,206 @@
+
+        /// <summary>
+        /// Whether or not the current BitPosition is evenly divisible by 8. I.e. whether or not the BitPosition is at a byte boundary.
+        /// </summary>
+        public bool BitAligned
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => (m_BitPosition & 7) == 0;
+        }
+
+        internal unsafe BitWriter(ref FastBufferWriter writer)
+        {
+            m_Writer = new Ref<FastBufferWriter>(ref writer);
+            m_BufferPointer = writer.BufferPointer + writer.PositionInternal;
+            m_Position = writer.PositionInternal;
+            m_BitPosition = 0;
+#if DEVELOPMENT_BUILD || UNITY_EDITOR
+            m_AllowedBitwiseWriteMark = (m_Writer.Value.AllowedWriteMark - m_Writer.Value.Position) * k_BitsPerByte;
+#endif
+        }
+
+        /// <summary>
+        /// Pads the written bit count to byte alignment and commits the write back to the writer
+        /// </summary>
+        public void Dispose()
+        {
+            var bytesWritten = m_BitPosition >> 3;
+            if (!BitAligned)
+            {
+                // Accounting for the partial write
+                ++bytesWritten;
+            }
+
+            m_Writer.Value.CommitBitwiseWrites(bytesWritten);
+        }
+
+        /// <summary>
+        /// Verifies the requested bit count can be written to the buffer.
+        /// This exists as a separate method to allow multiple bit writes to be bounds checked with a single call.
+        /// If it returns false, you may not write, and in editor and development builds, attempting to do so will
+        /// throw an exception. In release builds, attempting to do so will write to random memory addresses and cause
+        /// Bad Things(TM).
+        /// </summary>
+        /// <param name="bitCount">Number of bits you want to write, in total</param>
+        /// <returns>True if you can write, false if that would exceed buffer bounds</returns>
+        public unsafe bool VerifyCanWriteBits(int bitCount)
+        {
+            var newBitPosition = m_BitPosition + bitCount;
+            var totalBytesWrittenInBitwiseContext = newBitPosition >> 3;
+            if ((newBitPosition & 7) != 0)
+            {
+                // Accounting for the partial write
+                ++totalBytesWrittenInBitwiseContext;
+            }
+
+            if (m_Position + totalBytesWrittenInBitwiseContext > m_Writer.Value.CapacityInternal)
+            {
+                if (m_Writer.Value.CapacityInternal < m_Writer.Value.MaxCapacityInternal)
+                {
+                    m_Writer.Value.Grow();
+                    m_BufferPointer = m_Writer.Value.BufferPointer + m_Writer.Value.PositionInternal;
+                }
+                else
+                {
+                    return false;
+                }
+            }
+#if DEVELOPMENT_BUILD || UNITY_EDITOR
+            m_AllowedBitwiseWriteMark = newBitPosition;
```

Comment:
I find it a little odd that a `Verify` method (which sounds like a pure function) mutates state. Nothing in the XML comment suggests that state is modified either.

Reply:
The comment says:

```c#
        /// Unsafe write operations will throw OverflowException in editor and development builds if you
        /// go past the point you've marked using VerifyCanWrite(). In release builds, OverflowException will not be thrown
        /// for performance reasons, since the point of using VerifyCanWrite is to avoid bounds checking in the following
        /// operations in release builds.
```

Does that not indicate that VerifyCanWrite is going to mutate state in editor and development builds? In release builds, it doesn't... it only does it here as a safeguard to alert developers when they forgot to do bounds checking.

Reply:
Where is that text coming from?

Reply:
Also, I started this comment here but I see above that state is mutated in non-development contexts too (`m_BufferPointer = m_Writer.Value.BufferPointer + m_Writer.Value.PositionInternal;`)

Reply:
Oh, sorry, wrong Verify method... that was in FastbufferWriter. This one just says:

```c#
        /// If it returns false, you may not write, and in editor and development builds, attempting to do so will
        /// throw an exception.
```

I'll update this text to match the main reader/writers.

Reply:
Sounds good. I'd suggest workshopping the Verify name though. At face value, it screams const to me. I don't have any specific suggestions though and I'll leave it to your discretion.

Reply:
> Also, I started this comment here but I see above that state is mutated in non-development contexts too > >(m_BufferPointer = m_Writer.Value.BufferPointer + m_Writer.Value.PositionInternal;)

Write, also forgot to mention in the comment on this one that the buffer will grow if there's not enough space. That's what that part is. Instead of returning false and making you grow it yourself, it grows it for you to accommodate your request and then returns true.

That's done for performance reasons so you can do one big bounds check (and grow if necessary) before writing 20 fields, instead of writing 20 fields with 20 bounds checks slowing them all down.

Reply:
Perhaps `BeginWrite()` is a better name? But there's no corresponding `EndWrite()` so that might be confusing...

Reply:
I do like that, it informs that this should be called in sequence also it is pretty obvious that it will modify some state. Although it doesn't really describe the return value

Reply:
TryBeginWrite? :P

Reply:
Plenty of APIs have nop EndWrites. You could also do something like ReserveBytes which gets across the mutative behavior here, but not the required-order aspect.